src/MongoDB.Driver.Core/Core/Configuration/ConnectionString.cs

```diff
@@ -759,8 +759,14 @@ private void Parse()
                 @"((?<username>[^:@]+)(:(?<password>[^:@]*))?@)?" +
                 serversPattern + @"(/" + databasePattern + ")?/?" + optionsPattern + "$";
 
+            Match percentMatch = null;
+            if (_originalConnectionString.Contains("%"))
+            {
+                percentMatch = Regex.Match(_originalConnectionString, "(%[0-9A-F-a-f]{2})+");
```

Comment:
Glad to see the unit tests passing! I think there may a minor hiccup here though: consider the following string
`mongodb://user%20%:password@localhost`--I think this should be invalid?

One possible approach could be to check for the invalid combinations:
1. % followed a non hex character
2. % followed by a hex character and a non hex character
Off the cuff, I think this would loosely look like:
``(%[^0-9A-Fa-f])|(%[0-9A-Fa-f][^0-9A-Fa-f])``
You might also be able to use [string interpolation](https://docs.microsoft.com/en-us/dotnet/csharp/language-reference/tokens/interpolated) to make the regex vaguely easier to read by defining an intermediate value such as 
```
var hexDigit= "0-9A-Fa-f";
var pattern = $"(%[^{hexDigit}])|"  // % followed by a non-hexadecimal digit OR
    +$"(%[{hexDigit}][^{hexDigit}])"; // % followed by a hexadecimal digit and a non-hexadecimal digit
```